### ISSUE TITLE
Homebridge mqttthing securitySystem

### ISSUE THREAD

**Issue by marcantoineveilleux** [ParadoxAlarmInterface/pai#38]
Hi!

How should I configure my homebridge plugin when `MQTT_HOMEBRIDGE_ENABLE=True`?

I tried 

```javascript
 {
    "accessory": "mqttthing",
    "type": "securitySystem",
    "name": "Alarm",
    "url": "<url>",
    "username": "<user>",
    "password": "<password>",
    "topics": {
        "setTargetState": "paradox/control/partitions/Area_1/target",
        "getTargetState": "paradox/control/partitions/Area_1/current",
        "getCurrentState": "paradox/control/partitions/Area_1/current"
    },
    "targetStateValues": [
        "STAY_ARM",
        "AWAY_ARM",
        "NIGHT_ARM",
        "DISARM"
    ],
    "currentStateValues": [
        "STAY_ARM",
        "AWAY_ARM",
        "NIGHT_ARM",
        "DISARM",
        "Triggered"
    ]
}
```

But it seems like it doesn't work well when I manually enable/disable the alarm system.

Any clues? Maybe we should add an example in the wiki?

**Comment by marcantoineveilleux** [comment]
Thanks! Already tried. When using the `paradox/states/partitions/Area_1/current` for the currentState, there seem to be a problem since the targetState is by default `DISARM`. So when I enable my alarm system, in the Home app, it appears like "disarming". 

I've dig up the homebridge, nodejs-hab projects and apple spec. From what I understand right now, we need to publish to the control topic the state of the hardware and not only the homebridge stuff. E.G: when you arm/disarm with the keypad, I think we need to push the intent to the topic and retain the state.

I'll try a couple of things in the next days.

**Comment by jpbarraca** [comment]
I don't think I understand your conclusions.
You have to publish the desired state to paradox/control/partitions/PARTITION_NAME and the current state will be exposed to paradox/states/partitions/PARTITION_NAME/current.

**Comment by yozik04** [comment]
paradox/states/partitions/PARTITION_NAME/current will update when pai will receive an arm/disarm event.

**Comment by marcantoineveilleux** [comment]
Yes, agreed, `paradox/states/partitions/PARTITION_NAME/current` will always have the status of the system.

Let me try to explain again. I think the problem is the `setTargetState` and `getTargetState`. When configured like that:
```javascript
  "topics": {
                "setTargetState": "paradox/control/partitions/PARTITION_NAME",
                "getTargetState": "paradox/control/partitions/PARTITION_NAME",
                "getCurrentState": "paradox/states/partitions/PARTITION_NAME/current"
            },
```

By default, the target state is "disarm". So it means that if you arm physically your alarm, the `paradox/states/partitions/PARTITION_NAME/current` will become `armed`, but it won't update the `paradox/control/partitions/PARTITION_NAME`. Thus, homekit see the target state as `disarm`, with the current state at `armed`. So, if your system is armed and the target is to disarm, you are in the process of `disarming`. This is what I see in the Home app when I arm my system with the keypad.

**Comment by marcantoineveilleux** [comment]
It could work, but I am afraid of other kind of problems in edge cases. From the apple documentation about the [target security state](https://developer.apple.com/documentation/homekit/hmcharacteristicvaluetargetsecuritysystemstate) and the [current security state](https://developer.apple.com/documentation/homekit/hmcharacteristicvaluecurrentsecuritysystemstate), we can see that the number of constants and the contants themselves are not the same. (`stayArm`, `awayArm`, `nightArm`, `disarm` for the _target_ and `stayArm`, `awayArm`, `nightArm`, `disarmed`, `triggered` for the _current_ state)

**Comment by yozik04** [comment]
Homebridge is that NodeJS thing? How do you pair it with MQTT? Do you have a manual by hand. I would like to give it a try in the near future.

**Comment by marcantoineveilleux** [comment]
@yozik04 yes, that's the nodeJS thing.

To test fully, you need to have an iOS device and either an apple TV / iPad / homepod.

- Install npm
- Install [homebridge](https://github.com/nfarina/homebridge) via npm
- Install [mqttthings](https://github.com/arachnetech/homebridge-mqttthing) via npm
- [Configure your security system](https://github.com/arachnetech/homebridge-mqttthing#security-system) with the MQTT.
- Run PAI :)
- Restart homebridge and pair it with your Home using the Home app on your iOS device.
- Talam, it works!

**Comment by yozik04** [comment]
Did not know that iPad will also work. Will try, sounds to be very simple.

**Comment by marcantoineveilleux** [comment]
@yozik04 yep. Here are the instructions for the "apple part": https://www.imore.com/homekit-home-hub

**Comment by yozik04** [comment]
From: https://github.com/arachnetech/homebridge-mqttthing#security-system

```
"topics": {
    "setTargetState": "test/security/target",
    "getTargetState": "test/security/current",
    "getCurrentState": "test/security/current"
},
```

getTargetState and getCurrentState point to the same topic. Does it work this way for you?

**Comment by marcantoineveilleux** [comment]
No, but in their case it works because the values are the same (`Disarmed` == `Disarmed` and not `Disarm`)
```javascript
    "targetStateValues": [ "StayArm", "AwayArm", "NightArm", "Disarmed" ],
    "currentStateValues": [ "StayArm", "AwayArm", "NightArm", "Disarmed", "Triggered" ]
```

**Comment by yozik04** [comment]
They are also not consistent. Arm but Disarm**ed**.

```
"getTargetState": {
          "topic": "paradox/control/partitions/Area_1/current",
          "apply": "return message=='Disarm'?"Disarmed":message;"
      }
```

No?

**Comment by yozik04** [comment]
Tried this homebridge in docker. It could not distinguish between boiler temperature and room temperature. This is no go for me. Siri is also too dumb to understand when I say boiler or hot water...

**Comment by jpbarraca** [comment]
We should look at the documentation to understand the glossary. Could it be that one refers to a specific partition and and the other to the entire system?

**Comment by marcantoineveilleux** [comment]
oh! I didn't see the "apply" method. This could definitely work. We would lose the "disarming" or "arming" message (since both topic would always have the same value) but who cares! I'll try that.

**Comment by marcantoineveilleux** [comment]
seems like
```javascript
            "topics": {
                "setTargetState": "paradox/control/partitions/Area_1",
                "getTargetState": {
                    "topic": "paradox/states/partitions/Area_1/current",
                    "apply": "return message=='DISARMED'?'DISARM':message;"
                },
                "getCurrentState": "paradox/states/partitions/Area_1/current"
            },
            "targetStateValues": [
                "STAY_ARM",
                "AWAY_ARM",
                "NIGHT_ARM",
                "DISARM"
            ],
            "currentStateValues": [
                "STAY_ARM",
                "AWAY_ARM",
                "NIGHT_ARM",
                "DISARMED",
                "ALARM_TRIGGERED"
            ]
        },
```

Is a good solution. I'll test it more and probably make a PR for the wiki.

**Comment by jpbarraca** [comment]
From what I understand, the ```targetState``` represents the state desired by the user, while the ```currentState" represents the state that is actually active. 

An alternative solution would be to move the Homebridge topics to ```paradox/homebridge/PARTITION/get_current```, ```paradox/homebridge/PARTITION/get_target``` and ```paradox/homebridge/PARTITION/set_target```. The objective is to keep standard MQTT states and controls independent of Homebridge, facilitating maintenance and debug.

**Comment by iondarie** [comment]
If you are using: Homebridge plugin supporting various services over MQTT (https://github.com/arachnetech/homebridge-mqttthing#readme)

The following config works almost perfect for me:

`{
            "accessory": "mqttthing",
            "type": "securitySystem",
            "name": "Alarm P",
            "url": "mqtt://localhost",
            "topics": {
                "setTargetState": "paradox/control/partitions/Parter/",
                "getTargetState": "paradox/states/partitions/Parter/current",
                "getCurrentState": "paradox/states/partitions/Parter/current"
            },
            "currentStateValues": [
                "STAY_ARM",
                "AWAY_ARM",
                "NIGHT_ARM",
                "DISARMED",
                "ALARM_TRIGGERED"
            ],
            "targetStateValues": [
                "STAY_ARM",
                "AWAY_ARM",
                "NIGHT_ARM",
                "DISARM"
            ]
        }`

Parter is the name of my partition.

**Comment by jpbarraca** [comment]
Check #49 as the states and commands are now configurable.

**Comment by jpbarraca** [comment]
Is this still valid?

**Comment by iondarie** [comment]
I think this can be closed mqttthing works perfectly with PAI i am using it every single day.
